### PR TITLE
Refactor CompImageHDU to inherit from ImageHDU

### PR DESCRIPTION
**Pull request by astrofrog**
This pull request addresses https://github.com/astropy/astropy/issues/9238 by refactoring ``CompImageHDU`` to inherit from ``ImageHDU`` instead of ``BinTableHDU``. This includes some subtle API changes that cannot be warned about in advance using deprecation warnings or future warnings, so this PR, if accepted, would definitely warrant a major version bump. Ideally this would go in v6.0 but it depends on whether we can achieve consensus and stamp out remaining issues.

This PR isn't 100% ready (see 'Outstanding issues' below and FIXME comments in code) but I wanted to try and open this now in case it is something we can consider getting in before the v6.0 feature freeze. I doubt that the outstanding issues would cause significant code changes so it should be safe to already review things as-is.

### Background

Compressed image HDUs are stored in FITS files as binary tables. The ``CompImageHDU`` was therefore developed as a a subclass of ``BinTableHDU``. However, that implementation is quite complex because it tries to look to users like an ``ImageHDU`` - for example ``.header`` is meant to be the image header, and ``.data`` is meant to be the decompressed image data. To quote comments from the code:

```
# TODO: The difficulty of implementing this screams a need to rewrite this module
```

```
Bypasses `BinTableHDU._writeheader()` which updates the header with
metadata about the data that is meaningless here; another reason
why this class maybe shouldn't inherit directly from BinTableHDU...
```

Issue https://github.com/astropy/astropy/issues/9238 by @mhvk also suggested that ``CompImageHDU`` would be simpler if it inherited from ``ImageHDU``.

This PR does that refactoring, and represents ``CompImageHDU`` as an ``ImageHDU`` subclass that internally stores a ``BinTableHDU`` representing the compressed data. Instead of always keeping the image and table header in sync, headers are only converted at I/O time which should in principle make things faster.

Overall, this results in a few hundred lines of code being removed. For now ``CompImageHeader`` is kept to avoid breaking API so that warnings about reserved keywords are emitted as soon as they are set. However, if we would be happy to 'break' this behavior and only warn at I/O time or when calling ``verify``, we could get rid of the custom header subclass. If we want to keep the behavior, we can nevertheless try and avoid code duplication with the ``Header`` class by ssimplifying the header validation (see https://github.com/astropy/astropy/pull/15293).

### API changes/fixes

I have tried to minimize API changes as much as possible. However, it is impossible to not have any changes at all:

* With this PR, a compressed image HDU ``chdu`` will no longer return ``True`` for ``isinstance(chdu, BinTableHDU)``. There's no way to avoid this and also no way to do this with a gradual deprecation. However, I would hope that this is rare and that if anything people would do ``isinstance(chdu, CompImageHDU)`` which will still be ``True``.

* The following options have changed defaults in the ``CompImageHDU`` initializer to match ``ImageHDU``:

    * ``uint=True`` replaces ``uint=False`` (I need to check if I can come up with an example that shows the difference here)
    * ``scale_back=None`` replaces ``scale_back=False`` (in practice I think this won't change anything)

* The checksum on the main header is now the image checksum not the binary table checksum (I believe this was a bug: https://github.com/astropy/astropy/issues/14753)

### Modified tests

I've tried to change tests as little as possible but there were some unavoidable changes:

* Many tests (probably non-optimally) were checking the contents of ``chdu._header`` which no longer exists (the equivalent is ``chdu._bintable.header``.
* ``test_compressed_header_missing_znaxis`` has been removed - this checked that removing ``ZNAXIS`` or ``ZBITPIX`` from the internal binary table caused accessing the compressed data property to fail. This doesn't seem needed because it relies on modifying private attributes on ``CompImageHDU`` and we now also generate the compressed header on-the-fly so there is no opportunity to remove the keywords before the data is compressed.
* ``test_compressed_header_double_extname`` - included some lines checking state of internal/private properties. Those checks have been removed as the test is no longer relevant for this API.

### Outstanding issues/open questions

* [ ] Add back support for accessing ``CompImageHDU.shape`` (should be simple)
* [ ] Make sure that we don't write out anything when in ``mode='update'`` if a file is simply opened and closed (``test_open_comp_image_in_update_mode`` failure)
* [ ] Not critical for this PR, but currently this preserves the ``CompImageHeader`` class - as noted higher up we could potentially consider removing or simplifying this in future.
* [ ] This deprecates but does not yet remove the ``compressed_data`` property on ``CompImageHDU``. We could keep it but it is probably better to encourage people to open the file with ``disable_image_compression``? I doubt this property is used much.

### CODE REVIEW

File: astropy/io/fits/hdu/compressed.py [hunk 1]
```diff
@@ -675,58 +534,109 @@ def __init__(
...
-        self._bitpix = self._header["ZBITPIX"]
+            # TODO: just for parameter validation, e.g. tile shape - we shouldn't
+            # ideally need this and should instead validate the values as they are
+            # set above.
+            self._get_bintable_without_data()
 
-        self._orig_bzero = self._bzero
-        self._orig_bscale = self._bscale
-        self._orig_bitpix = self._bitpix
+    @property
```
Comment (mhvk) [inline astropy/io/fits/hdu/compressed.py:637]: Is this redefined to avoid inheriting the setter?

Comment (astrofrog) [reply astropy/io/fits/hdu/compressed.py:637]: I can't remember - but there's a lot of hacks/workarounds currently in this PR, I will go through systematically to make sure everything is needed and will add comments in places.

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 2]
```diff
@@ -526,6 +526,13 @@ def compress_image_data(
     for irow, tile_slices in _iter_array_tiles(data_shape, tile_shape):
         tile_data = image_data[tile_slices]
 
+        # FIXME: hack for now
```
Comment (Cadair) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:529]: :rofl:

**Comment by astrofrog** [comment]
Closing as the astropy PR is now open.

**Pull request closed without merging**